File: Conversoes de Bases/Decimal e Binario/DecimalBinario.java
import java.util.Scanner;

public class Main { // DecimalBinario

    private static final int LimiteMaximoDecimal = 2147483647;
    private static final String LimiteMaximoBinario = "1111111111111111111111111111111";

    private static String decimalParaBinario(int decimal) {
        StringBuilder binario = new StringBuilder();
        while (decimal > 0) {
            int resto = decimal % 2;
            binario.insert(0, resto);
            decimal /= 2;
        }
        while (binario.length() < 32) {
            binario.insert(0, "0");
        }
        return binario.toString();
    }

    private static int binarioParaDecimal(String binario) {
        int valor = 0;
        for (int i = binario.length() - 1; i >= 0; i--) {
            if (binario.charAt(i) == '1') {
                int potencia = binario.length() - 1 - i;
                valor += 1 << potencia;
            }
        }
        return valor;
    }

    private static boolean isValidBinario(String binario) {
        for (int i = 0; i < binario.length(); i++) {
            if (binario.charAt(i) != '0' && binario.charAt(i) != '1') {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int opcao = 0;
        int decimal;
        String binario;
        do {
            System.out.println("===========================");
            System.out.println(" CONVERSOR DECIMAL-BINARIO ");
            System.out.println("===========================");
            System.out.println();
            System.out.println("1 - Converter decimal para binario");
            System.out.println("2 - Converter binario para decimal");
            System.out.println("0 - Sair");
            System.out.println();
            System.out.print("Digite a opcao desejada: ");
            opcao = scanner.nextInt();
            switch (opcao) {
                case 1:
                    System.out.println();
                    System.out.print("Digite um numero decimal (ate " + LimiteMaximoDecimal + "): ");
                    decimal = scanner.nextInt();
                    if (decimal > LimiteMaximoDecimal) {
                        System.out.println("Valor decimal fora do limite permitido.");
                    } else {
                        String bin = decimalParaBinario(decimal);
                        System.out.println("Valor em binario: " + bin);
                    }
                    break;
                case 2:
                    System.out.println();
                    System.out.print("Digite um numero binario (ate " + LimiteMaximoBinario + "): ");
                    binario = scanner.next();
                    if (!isValidBinario(binario)) {
                        System.out.println("Valor binario invalido.");
                    } else if (binario.length() > LimiteMaximoBinario.length()) {
                        System.out.println("Valor binario fora do limite permitido.");
                    } else {
                        int dec = binarioParaDecimal(binario);
                        System.out.println("Valor em decimal: " + dec);
                    }
                    break;
                case 0:
                    break;
                default:
                    System.out.println("Opcao invalida. Tente novamente.");
                    break;
            }
            System.out.println();
        } while (opcao != 0);
        scanner.close();
    }
}

File: Conversoes de Bases/Decimal e Binario/DecimalBinario.scala
import scala.io.StdIn.readLine

object Main { // DecimalBinario

  val LimiteMaximoDecimal: Long = 2147483647
  val LimiteMaximoBinario: String = "1111111111111111111111111111111"

  def decimalParaBinario(decimal: Long): String = {
    var binario = ""
    var tempDecimal = decimal
    while (tempDecimal > 0) {
      val resto = (tempDecimal % 2).toString
      binario = resto + binario
      tempDecimal /= 2
    }
    while (binario.length < 32) {
      binario = "0" + binario
    }
    binario
  }

  def binarioParaDecimal(binario: String): Long = {
    var valor = 0L
    for (i <- binario.length - 1 to 0 by -1) {
      if (binario(i) == '1') {
        val potencia = binario.length - 1 - i
        valor += Math.pow(2, potencia).toLong
      }
    }
    valor
  }

  def isValidBinario(binario: String): Boolean = {
    binario.forall(c => c == '0' || c == '1')
  }

  def main(args: Array[String]): Unit = {
    var opcao = 0
    var decimal = 0L
    var binario = ""
    do {
      println("===========================")
      println(" CONVERSOR DECIMAL-BINARIO ")
      println("===========================")
      println()
      println("1 - Converter decimal para binario")
      println("2 - Converter binario para decimal")
      println("0 - Sair")
      println()
      print("Digite a opcao desejada: ")
      opcao = readLine().toInt
      opcao match {
        case 1 =>
          println()
          print(s"Digite um numero decimal (ate $LimiteMaximoDecimal): ")
          decimal = readLine().toLong
          if (decimal > LimiteMaximoDecimal)
            println("Valor decimal fora do limite permitido.")
          else
            println(s"Valor em binario: ${decimalParaBinario(decimal)}")
        case 2 =>
          println()
          print(s"Digite um numero binario (ate $LimiteMaximoBinario): ")
          binario = readLine()
          if (!isValidBinario(binario))
            println("Valor binario invalido.")
          else if (binario.length > LimiteMaximoBinario.length)
            println("Valor binario fora do limite permitido.")
          else
            println(s"Valor em decimal: ${binarioParaDecimal(binario)}")
        case 0 =>
        case _ => println("Opcao invalida. Tente novamente.")
      }
      println()
    } while (opcao != 0)
  }
}
